diff --git a/src/Builders/EnvironmentBuilder/JsonEnvironmentBuilder/JsonEnvironmentBuilder.cpp b/src/Builders/EnvironmentBuilder/JsonEnvironmentBuilder/JsonEnvironmentBuilder.cpp
index ca01940..1bf4bc8 100644
--- a/src/Builders/EnvironmentBuilder/JsonEnvironmentBuilder/JsonEnvironmentBuilder.cpp
+++ b/src/Builders/EnvironmentBuilder/JsonEnvironmentBuilder/JsonEnvironmentBuilder.cpp
@@ -200,9 +200,6 @@ std::unique_ptr<IRenderer> JsonEnvironmentBuilder::parseRenderer(
     std::unique_ptr<IProgressIndicator> progressIndicator =
         std::make_unique<CoutProgressIndicator>();
 
-    std::unique_ptr<IBackground> background =
-        parseBackground(jsonContent);
-
     const bool photonMappingEnabled =
         jsonContent["renderer"]["photonMappingEnabled"].get<bool>();
 
@@ -227,7 +224,6 @@ std::unique_ptr<IRenderer> JsonEnvironmentBuilder::parseRenderer(
             std::move(lightSampler),
             std::move(progressIndicator),
             std::move(photonMapBuilder),
-            std::move(background),
             parameters
         );
     }
@@ -244,7 +240,6 @@ std::unique_ptr<IRenderer> JsonEnvironmentBuilder::parseRenderer(
         logger,
         std::move(lightSampler),
         std::move(progressIndicator),
-        std::move(background),
         parameters
     );
 }
@@ -370,11 +365,16 @@ std::unique_ptr<Scene> JsonEnvironmentBuilder::parseScene(
 
     parseSceneLights(sceneLights, jsonContent);
 
+    std::unique_ptr<IBackground> background =
+        parseBackground(jsonContent);
+
     std::unique_ptr<IHittable> sceneRoot =
         bvhBuilder.build(std::move(objects));
 
     auto scene = std::make_unique<Scene>(
-        std::move(sceneRoot), std::move(sceneLights)
+        std::move(sceneRoot),
+        std::move(sceneLights),
+        std::move(background)
     );
 
     return std::move(scene);
diff --git a/src/Builders/MeshBuilder/ObjMeshBuilder/ObjMeshBuilder.cpp b/src/Builders/MeshBuilder/ObjMeshBuilder/ObjMeshBuilder.cpp
index a5f5b39..c7e502e 100644
--- a/src/Builders/MeshBuilder/ObjMeshBuilder/ObjMeshBuilder.cpp
+++ b/src/Builders/MeshBuilder/ObjMeshBuilder/ObjMeshBuilder.cpp
@@ -4,6 +4,7 @@
 #include "Builders/MeshBuilder/ObjMeshBuilder/MtlParameters.hpp"
 #include "Core/Color/Color.hpp"
 #include "Geometry/Hittable/Triangle/Triangle.hpp"
+#include "Geometry/Light/ILight.hpp"
 #include "Geometry/Light/TriangleAreaLight/TriangleAreaLight.hpp"
 #include "Geometry/Vertex.hpp"
 #include "Rendering/Material/DiffuseMaterial/DiffuseMaterial.hpp"
@@ -352,11 +353,7 @@ void ObjMeshBuilder::parseFace(
             fanIndexB, vertexBuffer, textureCoordsBuffer, normalsBuffer
         );
 
-        auto triangle = std::make_unique<Triangle>(
-            baseVertex, vertexA, vertexB, material
-        );
-
-        triangleBuffer.push_back(std::move(triangle));
+        std::shared_ptr<ILight> triangleLight = nullptr;
 
         if (material && not material->getEmission().isBlack()) {
             auto light = std::make_shared<TriangleAreaLight>(
@@ -366,8 +363,16 @@ void ObjMeshBuilder::parseFace(
                 material->getEmission()
             );
 
+            triangleLight = light;
+
             areaLightBuffer.push_back(std::move(light));
         }
+
+        auto triangle = std::make_unique<Triangle>(
+            baseVertex, vertexA, vertexB, material, triangleLight
+        );
+
+        triangleBuffer.push_back(std::move(triangle));
     }
 }
 
diff --git a/src/Core/Ray/Ray.hpp b/src/Core/Ray/Ray.hpp
index 320d216..8d2cc1e 100644
--- a/src/Core/Ray/Ray.hpp
+++ b/src/Core/Ray/Ray.hpp
@@ -11,7 +11,7 @@ private:
 
 public:
     Ray() = default;
-    Ray(const Point3<float>& origin, const Vector3<float>& direction);
+    Ray(const Point3f& origin, const Vector3f& direction);
 
     [[nodiscard]] const Point3f& getOrigin() const;
     [[nodiscard]] const Vector3f& getDirection() const;
diff --git a/src/Geometry/Hittable/HitData.hpp b/src/Geometry/Hittable/HitData.hpp
index 8a3677b..2066f38 100644
--- a/src/Geometry/Hittable/HitData.hpp
+++ b/src/Geometry/Hittable/HitData.hpp
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "Core/Math/Point.hpp"
+#include "Geometry/Light/ILight.hpp"
 
 #include <memory>
 
@@ -13,5 +14,6 @@ struct HitData {
     Point2f textureCoords {0.0F, 0.0F};
     Vector3f hitNormal {0.0, 0.0, 0.0};
     std::shared_ptr<IMaterial> material;
+    std::shared_ptr<ILight> light;
 };
 }
\ No newline at end of file
diff --git a/src/Geometry/Hittable/Triangle/Triangle.cpp b/src/Geometry/Hittable/Triangle/Triangle.cpp
index 1629bff..d9591c7 100644
--- a/src/Geometry/Hittable/Triangle/Triangle.cpp
+++ b/src/Geometry/Hittable/Triangle/Triangle.cpp
@@ -1,9 +1,9 @@
 #include "Triangle.hpp"
 
 #include "Core/Math/Interval.hpp"
-#include "Core/Math/Numeric.hpp"
 #include "Core/Math/Vector.hpp"
 #include "Geometry/BoundingVolume/AxisAlignedBoundingBox/AxisAlignedBoundingBox.hpp"
+#include "Geometry/Light/ILight.hpp"
 #include "Rendering/Material/IMaterial.hpp"
 
 #include <memory>
@@ -41,58 +41,6 @@ Point2f Triangle::createTextureCoords(
            result.barycentricWeightC * vertexC_.texturePosition.value();
 }
 
-Triangle::MollerTrumboreResult Triangle::solveMollerTrumbore(
-    const Ray& ray
-) const {
-    const Vector3f reversedDirection = -ray.getDirection();
-    const Vector3f crossedEdge2 =
-        getCrossProduct(edge2_, reversedDirection);
-    const Vector3f solution = ray.getOrigin() - vertexC_.position;
-
-    const Interval<float> unitInterval {0.0F, 1.0F};
-    const float mainDeterminant = getDotProduct(edge1_, crossedEdge2);
-
-    if (isCloseToZero(mainDeterminant)) {
-        return MollerTrumboreResult {};
-    }
-
-    const float weight1Determinant =
-        getDotProduct(solution, crossedEdge2);
-
-    const float weight1 = weight1Determinant / mainDeterminant;
-
-    if (not unitInterval.contains(weight1)) {
-        return MollerTrumboreResult {};
-    }
-
-    const float weight2Determinant = getDotProduct(
-        edge1_, getCrossProduct(solution, reversedDirection)
-    );
-
-    const float weight2 = weight2Determinant / mainDeterminant;
-
-    if (not unitInterval.contains(weight2)) {
-        return MollerTrumboreResult {};
-    }
-
-    if (weight1 + weight2 > 1) {
-        return MollerTrumboreResult {};
-    }
-
-    const float rayTDeterminant =
-        getDotProduct(edge1_, getCrossProduct(edge2_, solution));
-
-    const float rayT = rayTDeterminant / mainDeterminant;
-
-    return MollerTrumboreResult {
-        .hasSolution = true,
-        .t0 = rayT,
-        .barycentricWeightA = weight1,
-        .barycentricWeightB = weight2,
-        .barycentricWeightC = 1.0F - weight1 - weight2
-    };
-}
-
 [[nodiscard]] AxisAlignedBoundingBox Triangle::createBoundingBox(
     const Point3f& vertexA,
     const Point3f& vertexB,
@@ -135,13 +83,15 @@ void Triangle::updateHitData(
     hitData.hitNormal = normal;
     hitData.textureCoords = textureCoords;
     hitData.material = material_;
+    hitData.light = light_;
 }
 
 Triangle::Triangle(
     Vertex vertexA,
     Vertex vertexB,
     Vertex vertexC,
-    std::shared_ptr<IMaterial> material
+    std::shared_ptr<IMaterial> material,
+    std::shared_ptr<ILight> light
 ) :
     vertexA_(vertexA),
     vertexB_(vertexB),
@@ -152,6 +102,7 @@ Triangle::Triangle(
         vertexC_.position
     )),
     material_(std::move(material)),
+    light_(std::move(light)),
     edge1_(vertexA_.position - vertexC_.position),
     edge2_(vertexB_.position - vertexC_.position),
     flatNormal_(createFlatNormalIfNeccessary()) {}
@@ -166,7 +117,9 @@ bool Triangle::hitClosest(
     const Interval<float>& interval,
     HitData& hitData
 ) const {
-    const MollerTrumboreResult result = solveMollerTrumbore(ray);
+    const MollerTrumboreResult result = intersectTriangle(
+        ray, vertexA_.position, vertexB_.position, vertexC_.position
+    );
 
     if (not result.hasSolution) {
         return false;
@@ -184,7 +137,9 @@ bool Triangle::hitAny(
     const Ray& ray,
     const Interval<float>& interval
 ) const {
-    const MollerTrumboreResult result = solveMollerTrumbore(ray);
+    const MollerTrumboreResult result = intersectTriangle(
+        ray, vertexA_.position, vertexB_.position, vertexC_.position
+    );
 
     return interval.contains(result.t0);
 }
diff --git a/src/Geometry/Hittable/Triangle/Triangle.hpp b/src/Geometry/Hittable/Triangle/Triangle.hpp
index 6e7a880..aff73a9 100644
--- a/src/Geometry/Hittable/Triangle/Triangle.hpp
+++ b/src/Geometry/Hittable/Triangle/Triangle.hpp
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "Core/Math/Intersections.hpp"
 #include "Geometry/BoundingVolume/AxisAlignedBoundingBox/AxisAlignedBoundingBox.hpp"
 #include "Geometry/Hittable/IHittable.hpp"
 #include "Geometry/Vertex.hpp"
@@ -16,19 +17,12 @@ private:
 
     const AxisAlignedBoundingBox boundingBox_;
     std::shared_ptr<IMaterial> material_;
+    std::shared_ptr<ILight> light_ = nullptr;
 
     const Vector3f edge1_;
     const Vector3f edge2_;
     const std::optional<Vector3f> flatNormal_;
 
-    struct MollerTrumboreResult {
-        bool hasSolution = false;
-        float t0 = 0.0F;
-        float barycentricWeightA = 0.0F;
-        float barycentricWeightB = 0.0F;
-        float barycentricWeightC = 0.0F;
-    };
-
     [[nodiscard]] std::optional<Vector3f>
     createFlatNormalIfNeccessary() const;
 
@@ -40,10 +34,6 @@ private:
         const MollerTrumboreResult& result
     ) const;
 
-    [[nodiscard]] MollerTrumboreResult solveMollerTrumbore(
-        const Ray& ray
-    ) const;
-
     [[nodiscard]] AxisAlignedBoundingBox createBoundingBox(
         const Point3f& vertexA,
         const Point3f& vertexB,
@@ -63,7 +53,8 @@ public:
         Vertex vertexA,
         Vertex vertexB,
         Vertex vertexC,
-        std::shared_ptr<IMaterial> material
+        std::shared_ptr<IMaterial> material,
+        std::shared_ptr<ILight> light = nullptr
     );
 
     [[nodiscard]] const AxisAlignedBoundingBox&
diff --git a/src/Geometry/Light/DirectionalLight/DirectionalLight.cpp b/src/Geometry/Light/DirectionalLight/DirectionalLight.cpp
index 82c7eb7..873e619 100644
--- a/src/Geometry/Light/DirectionalLight/DirectionalLight.cpp
+++ b/src/Geometry/Light/DirectionalLight/DirectionalLight.cpp
@@ -86,4 +86,11 @@ Photon DirectionalLight::emitPhoton() const {
         .position = randomPoint, .direction = direction_, .power = power_
     };
 }
+
+float DirectionalLight::calculatePdf(
+    const Point3f& origin,
+    const Vector3f& inDirection
+) const {
+    return 0.0F;
+}
 }
\ No newline at end of file
diff --git a/src/Geometry/Light/DirectionalLight/DirectionalLight.hpp b/src/Geometry/Light/DirectionalLight/DirectionalLight.hpp
index 067a4fc..da4a8be 100644
--- a/src/Geometry/Light/DirectionalLight/DirectionalLight.hpp
+++ b/src/Geometry/Light/DirectionalLight/DirectionalLight.hpp
@@ -41,5 +41,10 @@ public:
     [[nodiscard]] LinearColor getPower() const override;
 
     [[nodiscard]] Photon emitPhoton() const override;
+
+    [[nodiscard]] float calculatePdf(
+        const Point3f& origin,
+        const Vector3f& inDirection
+    ) const override;
 };
 }
\ No newline at end of file
diff --git a/src/Geometry/Light/ILight.hpp b/src/Geometry/Light/ILight.hpp
index d804d71..ba3311f 100644
--- a/src/Geometry/Light/ILight.hpp
+++ b/src/Geometry/Light/ILight.hpp
@@ -42,5 +42,10 @@ public:
     [[nodiscard]] virtual LinearColor getPower() const = 0;
 
     [[nodiscard]] virtual Photon emitPhoton() const = 0;
+
+    [[nodiscard]] virtual float calculatePdf(
+        const Point3f& origin,
+        const Vector3f& inDirection
+    ) const = 0;
 };
 }
\ No newline at end of file
diff --git a/src/Geometry/Light/PointLight/PointLight.cpp b/src/Geometry/Light/PointLight/PointLight.cpp
index ade0038..5ac688a 100644
--- a/src/Geometry/Light/PointLight/PointLight.cpp
+++ b/src/Geometry/Light/PointLight/PointLight.cpp
@@ -63,4 +63,11 @@ Photon PointLight::emitPhoton() const {
         .power = emission_
     };
 }
+
+float PointLight::calculatePdf(
+    const Point3f& origin,
+    const Vector3f& inDirection
+) const {
+    return 0.0F;
+}
 }
\ No newline at end of file
diff --git a/src/Geometry/Light/PointLight/PointLight.hpp b/src/Geometry/Light/PointLight/PointLight.hpp
index 4f17504..b7bf7a0 100644
--- a/src/Geometry/Light/PointLight/PointLight.hpp
+++ b/src/Geometry/Light/PointLight/PointLight.hpp
@@ -38,5 +38,10 @@ public:
     [[nodiscard]] LinearColor getPower() const override;
 
     [[nodiscard]] Photon emitPhoton() const override;
+
+    [[nodiscard]] float calculatePdf(
+        const Point3f& origin,
+        const Vector3f& inDirection
+    ) const override;
 };
 }
\ No newline at end of file
diff --git a/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.cpp b/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.cpp
index 952fbc8..2fce9d6 100644
--- a/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.cpp
+++ b/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.cpp
@@ -1,6 +1,7 @@
 #include "TriangleAreaLight.hpp"
 
 #include "Core/Color/Color.hpp"
+#include "Core/Math/Intersections.hpp"
 #include "Core/Math/Random.hpp"
 #include "Core/Math/Sampling.hpp"
 #include "Core/Math/Transformations.hpp"
@@ -88,7 +89,8 @@ void TriangleAreaLight::discretize(
     std::vector<std::shared_ptr<ILight>>& discreteLights,
     uint32_t samples
 ) const {
-    const LinearColor sampleEmission = (emission_ * area_) / float(samples);
+    const LinearColor sampleEmission =
+        (emission_ * area_) / float(samples);
 
     for (uint32_t i = 0; i < samples; i++) {
         const Point3f randomPoint = getRandomPoint();
@@ -148,4 +150,27 @@ Photon TriangleAreaLight::emitPhoton() const {
         .power = power_
     };
 }
+
+float TriangleAreaLight::calculatePdf(
+    const Point3f& origin,
+    const Vector3f& inDirection
+) const {
+    const Ray ray {origin, inDirection};
+
+    const MollerTrumboreResult result =
+        intersectTriangle(ray, vertexA_, vertexB_, vertexC_);
+
+    if (not result.hasSolution) {
+        return 0.0F;
+    }
+
+    const Point3f hitPoint = ray.at(result.t0);
+    const Vector3f toLight = hitPoint - origin;
+
+    const float distanceSquared = toLight.getSquaredLength();
+    const float cosine = getDotProduct(normal_, -inDirection);
+    const float solidAreaPdf = distanceSquared / (cosine * area_);
+
+    return solidAreaPdf;
+}
 }
\ No newline at end of file
diff --git a/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.hpp b/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.hpp
index 3419f5c..89c36e8 100644
--- a/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.hpp
+++ b/src/Geometry/Light/TriangleAreaLight/TriangleAreaLight.hpp
@@ -58,5 +58,10 @@ public:
     [[nodiscard]] LinearColor getPower() const override;
 
     [[nodiscard]] Photon emitPhoton() const override;
+
+    [[nodiscard]] float calculatePdf(
+        const Point3f& origin,
+        const Vector3f& inDirection
+    ) const override;
 };
 }
\ No newline at end of file
diff --git a/src/Rendering/DirectLightEstimator/DirectLightEstimator.cpp b/src/Rendering/DirectLightEstimator/DirectLightEstimator.cpp
index 427751d..7ee3da4 100644
--- a/src/Rendering/DirectLightEstimator/DirectLightEstimator.cpp
+++ b/src/Rendering/DirectLightEstimator/DirectLightEstimator.cpp
@@ -2,7 +2,9 @@
 
 #include "Core/Color/Color.hpp"
 #include "Core/Math/Numeric.hpp"
+#include "Geometry/Light/ILight.hpp"
 #include "Rendering/Material/IMaterial.hpp"
+#include "Rendering/Material/MaterialSample.hpp"
 #include "World/Scene/Scene.hpp"
 
 namespace RTC {
@@ -22,6 +24,86 @@ bool DirectLightEstimator::isInShadow(
     return scene.hitAny(shadowRay, interval);
 }
 
+LinearColor DirectLightEstimator::getLightRadiance(
+    const LightSample& lightSample,
+    const Scene& scene,
+    const HitData& hitData,
+    const Point3f& offsetHitPoint,
+    const Vector3f& outDirection
+) const {
+    if (isInShadow(offsetHitPoint, lightSample.toLight, scene)) {
+        return LinearColor::black();
+    }
+
+    const LinearColor brdf = hitData.material->calculateBrdf(
+        hitData, outDirection, lightSample.inDirection
+    );
+
+    const float cosine = std::max(
+        0.0F, getDotProduct(hitData.hitNormal, lightSample.inDirection)
+    );
+
+    const LinearColor lightRadiance =
+        brdf * cosine * lightSample.outLight;
+
+    const float lightPdf = lightSample.pdf;
+    const float materialPdf = hitData.material->calculatePdf(
+        hitData.hitNormal, lightSample.inDirection, outDirection
+    );
+
+    return lightRadiance / (lightPdf + materialPdf);
+}
+
+LinearColor DirectLightEstimator::getMaterialRadiance(
+    const ILight& sampledLight,
+    const MaterialSample& materialSample,
+    const Scene& scene,
+    const HitData& hitData,
+    const Point3f& offsetHitPoint,
+    const Vector3f& outDirection
+) const {
+    constexpr Interval<float> renderInterval {
+        EPSILON, Interval<float>::infinity()
+    };
+
+    const Ray scatterRay {offsetHitPoint, materialSample.inDirection};
+
+    HitData scatteredHitData;
+    const bool hitAnything =
+        scene.hitClosest(scatterRay, renderInterval, scatteredHitData);
+
+    if (not hitAnything) {
+        return scene.sampleBackground(scatterRay);
+    }
+
+    if (hitData.light.get() != &sampledLight) {
+        return BLACK_LINEAR_COLOR;
+    }
+
+    const LinearColor materialEmission =
+        scatteredHitData.material->getEmission(
+            scatteredHitData.hitPoint, -materialSample.inDirection
+        );
+
+    if (materialEmission.isBlack()) {
+        return BLACK_LINEAR_COLOR;
+    };
+
+    const float cosine = std::max(
+        0.0F, getDotProduct(hitData.hitNormal, materialSample.inDirection)
+    );
+
+    const LinearColor materialRadiance =
+        materialSample.brdf * cosine * materialEmission;
+
+    const float materialPdf = materialSample.pdf;
+    const float lightPdf = sampledLight.calculatePdf(
+        offsetHitPoint, materialSample.inDirection
+    );
+
+    return materialRadiance / (materialPdf + lightPdf);
+}
+
 LinearColor DirectLightEstimator::getScatteringWeight(
     const LightSample& lightSample,
     const Scene& scene,
@@ -53,10 +135,22 @@ LinearColor DirectLightEstimator::estimateRadiance(
 ) const {
     const LightSample lightSample = light.getSample(hitData.hitPoint);
 
-    const LinearColor scatteringWeight = getScatteringWeight(
+    const MaterialSample materialSample =
+        hitData.material->getSample(hitData, outDirection);
+
+    const LinearColor lightRadiance = getLightRadiance(
         lightSample, scene, hitData, offsetHitPoint, outDirection
     );
 
-    return scatteringWeight * lightSample.outLight;
+    const LinearColor materialRadiance = getMaterialRadiance(
+        light,
+        materialSample,
+        scene,
+        hitData,
+        offsetHitPoint,
+        outDirection
+    );
+
+    return lightRadiance + materialRadiance;
 }
 }
\ No newline at end of file
diff --git a/src/Rendering/DirectLightEstimator/DirectLightEstimator.hpp b/src/Rendering/DirectLightEstimator/DirectLightEstimator.hpp
index f01fc38..7ffc6f1 100644
--- a/src/Rendering/DirectLightEstimator/DirectLightEstimator.hpp
+++ b/src/Rendering/DirectLightEstimator/DirectLightEstimator.hpp
@@ -2,6 +2,7 @@
 
 #include "Geometry/Light/LightSample.hpp"
 #include "Rendering/DirectLightEstimator/IDirectLightEstimator.hpp"
+#include "Rendering/Material/MaterialSample.hpp"
 
 namespace RTC {
 class DirectLightEstimator : public IDirectLightEstimator {
@@ -12,6 +13,23 @@ private:
         const Scene& scene
     ) const;
 
+    [[nodiscard]] LinearColor getLightRadiance(
+        const LightSample& lightSample,
+        const Scene& scene,
+        const HitData& hitData,
+        const Point3f& offsetHitPoint,
+        const Vector3f& outDirection
+    ) const;
+
+    [[nodiscard]] LinearColor getMaterialRadiance(
+        const ILight& sampledLight,
+        const MaterialSample& materialSample,
+        const Scene& scene,
+        const HitData& hitData,
+        const Point3f& offsetHitPoint,
+        const Vector3f& outDirection
+    ) const;
+
 public:
     [[nodiscard]] LinearColor getScatteringWeight(
         const LightSample& lightSample,
diff --git a/src/Rendering/Renderer/PathRenderer/PathRenderer.cpp b/src/Rendering/Renderer/PathRenderer/PathRenderer.cpp
index f5fbf50..5b6d02e 100644
--- a/src/Rendering/Renderer/PathRenderer/PathRenderer.cpp
+++ b/src/Rendering/Renderer/PathRenderer/PathRenderer.cpp
@@ -20,7 +20,6 @@
 #include <utility>
 
 namespace RTC {
-
 LinearColor PathRenderer::getEmission(
     const HitData& hitData,
     const Vector3f& outDirection,
@@ -93,7 +92,7 @@ LinearColor PathRenderer::traceRay(
         scene.hitClosest(ray, renderInterval, hitData);
 
     if (not hitAnything) {
-        return background_->sample(ray);
+        return scene.sampleBackground(ray);
     }
 
     const Vector3f outDirection = -ray.getDirection().getNormalized();
@@ -227,13 +226,11 @@ PathRenderer::PathRenderer(
     std::shared_ptr<ILogger> logger,
     std::unique_ptr<ILightSampler> lightSampler,
     std::unique_ptr<IProgressIndicator> progressIndicator,
-    std::unique_ptr<IBackground> background,
     PathRendererParameters parameters
 ) :
     logger_(std::move(logger)),
     lightSampler_(std::move(lightSampler)),
     progressIndicator_(std::move(progressIndicator)),
-    background_(std::move(background)),
     parameters_(parameters) {}
 
 RendererStatistics PathRenderer::render(
diff --git a/src/Rendering/Renderer/PathRenderer/PathRenderer.hpp b/src/Rendering/Renderer/PathRenderer/PathRenderer.hpp
index 050ba8a..d48259b 100644
--- a/src/Rendering/Renderer/PathRenderer/PathRenderer.hpp
+++ b/src/Rendering/Renderer/PathRenderer/PathRenderer.hpp
@@ -22,7 +22,6 @@ private:
     std::shared_ptr<ILogger> logger_;
     std::unique_ptr<ILightSampler> lightSampler_;
     std::unique_ptr<IProgressIndicator> progressIndicator_;
-    std::unique_ptr<IBackground> background_;
     PathRendererParameters parameters_;
 
     [[nodiscard]] LinearColor getEmission(
@@ -67,7 +66,6 @@ public:
         std::shared_ptr<ILogger> logger,
         std::unique_ptr<ILightSampler> lightSampler,
         std::unique_ptr<IProgressIndicator> progressIndicator,
-        std::unique_ptr<IBackground> background,
         PathRendererParameters parameters
     );
 
diff --git a/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.cpp b/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.cpp
index db852f0..cdc5fd8 100644
--- a/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.cpp
+++ b/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.cpp
@@ -227,7 +227,7 @@ LinearColor PhotonMapRenderer::traceRay(
         scene.hitClosest(ray, renderInterval, hitData);
 
     if (not hitAnything) {
-        return background_->sample(ray);
+        return scene.sampleBackground(ray);
     }
 
     const Vector3f outDirection = -ray.getDirection().getNormalized();
@@ -366,14 +366,12 @@ PhotonMapRenderer::PhotonMapRenderer(
     std::unique_ptr<ILightSampler> lightSampler,
     std::unique_ptr<IProgressIndicator> progressIndicator,
     std::unique_ptr<IPhotonMapBuilder> photonMapBuilder,
-    std::unique_ptr<IBackground> background,
     PhotonMapRendererParameters parameters
 ) :
     logger_(std::move(logger)),
     lightSampler_(std::move(lightSampler)),
     progressIndicator_(std::move(progressIndicator)),
     photonMapBuilder_(std::move(photonMapBuilder)),
-    background_(std::move(background)),
     parameters_(parameters) {}
 
 RendererStatistics PhotonMapRenderer::render(
diff --git a/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.hpp b/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.hpp
index d566d07..e87e8d2 100644
--- a/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.hpp
+++ b/src/Rendering/Renderer/PhotonMapRenderer/PhotonMapRenderer.hpp
@@ -24,7 +24,6 @@ private:
     std::unique_ptr<ILightSampler> lightSampler_;
     std::unique_ptr<IProgressIndicator> progressIndicator_;
     std::unique_ptr<IPhotonMapBuilder> photonMapBuilder_;
-    std::unique_ptr<IBackground> background_;
     PhotonMapRendererParameters parameters_;
 
     struct PhotonMaps {
@@ -95,7 +94,6 @@ public:
         std::unique_ptr<ILightSampler> lightSampler,
         std::unique_ptr<IProgressIndicator> progressIndicator,
         std::unique_ptr<IPhotonMapBuilder> photonMapBuilder,
-        std::unique_ptr<IBackground> background,
         PhotonMapRendererParameters parameters
     );
 
diff --git a/src/World/Scene/Scene.cpp b/src/World/Scene/Scene.cpp
index 7f6de3a..f9cc21d 100644
--- a/src/World/Scene/Scene.cpp
+++ b/src/World/Scene/Scene.cpp
@@ -8,10 +8,12 @@
 namespace RTC {
 Scene::Scene(
     std::unique_ptr<IHittable> sceneRoot,
-    std::vector<std::shared_ptr<ILight>> lights
+    std::vector<std::shared_ptr<ILight>> lights,
+    std::unique_ptr<IBackground> background_
 ) :
     sceneRoot_(std::move(sceneRoot)),
-    lights_(std::move(lights)) {}
+    lights_(std::move(lights)),
+    background_(std::move(background_)) {}
 
 const std::vector<std::shared_ptr<ILight>>& Scene::getLights() const {
     return lights_;
@@ -26,6 +28,10 @@ void Scene::setup() {
     }
 }
 
+LinearColor Scene::sampleBackground(const Ray& ray) const {
+    return background_->sample(ray);
+}
+
 bool Scene::hitClosest(
     const Ray& ray,
     const Interval<float>& interval,
diff --git a/src/World/Scene/Scene.hpp b/src/World/Scene/Scene.hpp
index e85d77a..2096e02 100644
--- a/src/World/Scene/Scene.hpp
+++ b/src/World/Scene/Scene.hpp
@@ -2,7 +2,7 @@
 
 #include "Core/Math/Interval.hpp"
 #include "Core/Ray/Ray.hpp"
-#include "Geometry/BoundingVolume/AxisAlignedBoundingBox/AxisAlignedBoundingBox.hpp"
+#include "Geometry/Background/IBackground.hpp"
 #include "Geometry/Hittable/HitData.hpp"
 #include "Geometry/Hittable/IHittable.hpp"
 #include "Geometry/Light/ILight.hpp"
@@ -15,11 +15,13 @@ class Scene {
 private:
     std::unique_ptr<IHittable> sceneRoot_;
     std::vector<std::shared_ptr<ILight>> lights_;
+    std::unique_ptr<IBackground> background_;
 
 public:
     Scene(
         std::unique_ptr<IHittable> sceneRoot,
-        std::vector<std::shared_ptr<ILight>> lights
+        std::vector<std::shared_ptr<ILight>> lights,
+        std::unique_ptr<IBackground> background
     );
 
     void setup();
@@ -27,6 +29,8 @@ public:
     [[nodiscard]] const std::vector<std::shared_ptr<ILight>>&
     getLights() const;
 
+    [[nodiscard]] LinearColor sampleBackground(const Ray& ray) const;
+
     bool hitClosest(
         const Ray& ray,
         const Interval<float>& interval,
